src/lua/uri.lua

```diff
@@ -258,31 +258,45 @@ local function unescape(buf, opts)
     return str
 end
 
-local function encode_kv(key, values, res)
+local function encode_kv(key, values, res, escape_opts)
     local val = values
     if type(val) ~= "table" then
         val = { val }
     end
 
+    local key_escaped
+    if escape_opts ~= nil then
+        key_escaped = escape(key, escape_opts)
+    end
     -- { a = {} } --> "a"
     if next(val) == nil then
-        table.insert(res, key)
+        if escape_opts ~= nil then
+            table.insert(res, key_escaped)
+        else
+            table.insert(res, key)
+        end
```

Comment:
If we'll add a capture

```lua
-- Unlike escape() it doesn't perform escaping by default.
--
-- It is needed to <...>.
local function _escape(str)
    if escape_opts == nil then
        return str
    end
    return escape(str, escape_opts)
end
```

those if-else blocks will gone. They will be just:

```lua
table.insert(res, _escape(key))
<...>
local kv = ("%s=%s"):format(_escape(key), _escape(v))
```

It makes the code more straightforward.

However I don't see a reason why don't call just `escape()`. The functions of the question are internal, we don't need to keep any compatibility here.